ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1421 @@
+* the `WorkflowInstanceCreationIntent:Create` command creates the new workflow
+  instance
+* the `Process:Activate_Element` command transitions the process element to
+  activated
+* the `StartEvent:Activate_Element` command transitions the start event to
+  activated
+* the `StartEvent:Complete_Element` command transitions the start event to
+  completed and marks the sequence flow as taken
+
+On replay:
+* the `WorkflowInstanceCreationIntent:Created` event creates the new workflow
+  instance
+* the `Process:Element_Activating` event transitions the process element to
+  activating
+* the `Process:Element_Activated` event transitions the process element to
+  activated
+* the `StartEvent:Element_Activating` event transitions the start event to
+  activating
+* the `StartEvent:Element_Activated` event transitions the start event to
+  activated
+* the `StartEvent:Element_Completing` event transitions the start event to
+  completing
+* the `StartEvent:Element_Completed` event transitions the start event to
+  completed
+* the `Sequence_Flow_Taken` event marks the sequence flow as taken
+* processing can continue again at `ServiceTask:Activate_Element`
+
+Differences to current processing:
+* new commands for the state transitions of an element: `Activate_Element` and `Complete_Element`
+* the `...:Element_Activated`, `...:Element_Completed` and `Sequence_Flow_Taken`
+  events no longer produce follow-up records, because they are now strictly
+  events
+
+### Service Task
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+ServiceTask:Activate_Element | ServiceTask:Element_Activating, [Variable:Created], ServiceTask:Element_Activated, Job:Created
+JobBatch:Activate | JobBatch:Activated 
+Job:Complete | Job:Completed, ServiceTask:Complete_Element
+ServiceTask:Complete_Element | ServiceTask:Element_Completing, [Variable:Created], [Variable:Updated], ServiceTask:Element_Completed, Sequence_Flow_Taken, EndEvent:Activate_Element,
+
+Notes:
+* if the service task has an input variable mapping then it produces one or more `Variable:Created` events for the new local variables
+* if the job is completed with variables then these variables will be stored in the state until they are used for the output mapping (i.e. temporal/transient variables). They must therefore also be written into the `Job:Completed` event.
+* if the service task has an output variable mapping, or the job is completed with variables, then it produces one or more `Variable:Created`/`Variable:Updated` events for the new and updated variables of the workflow instance or scope
+* the order of writing follow-up events matters. Both `ServiceTask:Element_Activated` and `Variable:Created` need to be replayed before `Job:Created`.
+
+On processing:
+* the `ServiceTask:Activate_Element` command transitions the service task
+  element to activated, may create variables for input mapping and creates a job
+* the `JobBatch:Activate` command activates a batch of jobs
```

Comment:
You're right. It's requested/written by the gateway.

About distinguishing responses and state changes: We've tried to only put the state changes in the `On Processing` and `On replay` parts, since that is the part hit by this proposal. Side-effects (like responding to a client) are not affected, since that was already only on commands (AFAIK).